airflow/providers/grafana/log/loki_task_handler.py

```diff
@@ -0,0 +1,242 @@
+            str(task_instance.execution_date.isoformat()),
+        )
+        next_try = task_instance.next_try_number
+        start_epoch_time = self.convert_epoch_time(start_time)
+
+        if end_time and next_try < 3:
+            end_epoch_time = self.convert_epoch_time(end_time.isoformat())
+            return (start_epoch_time - 60, end_epoch_time + 60 * 2)
+        if end_time:
+            end_epoch_time = self.convert_epoch_time(end_time.isoformat())
+            execution_time_epoch_time = self.convert_epoch_time(execution_time)
+            return (execution_time_epoch_time - 60, end_epoch_time + 60 * 2)
+        return (start_epoch_time - 60, start_epoch_time + 60 * 60 * 24)
+
+    def get_read_parameters(self, task_instance, try_number) -> Dict[str, Union[str, int]]:
+        """
+        Construct query(LogQL) and update parameters for loki read request.
+        :param ti: task instance object
+        :type: task_instance: TaskInstance
+        :param try_number: current try_number to read log from
+        :type: try_number: int
+        """
+        start_time, end_time = self.query_time_range(task_instance)
+        task_lables = self.get_label(task_instance)
+        del task_lables["airflow_try_number"]  # try_number will come from read method
+
+        # converting task_lables(dict) to LogQL format
+        # {"dag_id":"airflow"} --> {dag_id = "airflow"}
+        query = "{"
+        for key, value in task_lables.items():
+            query += f"""{key}="{value}", """
+        query += f"""airflow_try_number="{try_number}"}} != "WARNING" """
+
+        return {
+            "direction": "forward",
+            "start": start_time,
+            "end": end_time,
+            "query": query,
+            "limit": 5000,
+        }
+
+    def _read(self, task_instance, try_number, metadata=None) -> Tuple[str, Dict[str, bool]]:
+        """
+        Return list of logs string for each task try number by streaming it from Loki
+        :param ti: task instance object
+        :type: task_instance: TaskInstance
+        :param try_number: current try_number to read log from
+        :type: try_number: int
+        """
+        logs = ""
+        if self.is_loki_alive() == 200:
+            query_parameters = self.get_read_parameters(task_instance, try_number)
+            logs += f"*** Loki Query:{query_parameters['query']} \n"
+            logs += (
+                f"*** Epoch start_time: {query_parameters['start']} "
+                f"and end_time {query_parameters['end']}\n"
+            )
+            try:
+                response = self.session.get(
+                    f"{self.get_conn.host}/loki/api/v1/query_range",
+                    params=query_parameters,
+                ).json()
+                results = response["data"]["result"]
+                logs_list = [value for result in results for value in result["values"]]
+                # sorting with list of values with timestamp.
```

Comment:
Why is this sorting needed? Is it in case multiple streams are returned? Because within one stream the result is already sorted when returned from loki.

https://grafana.com/docs/loki/latest/api/#query-loki-over-a-range-of-time

Reply:
In case of multiple streams, i believe this sorting can be done more efficiently as well, since each individual stream is already sorted. 
Instead of putting all values together into one logs_list and then sorting the whole list, you can iterate one item at a time from each of the value-lists, but only pick the one with highest timestamp when appending to the final log list. This should be O(N) rather than O(N logN) and one less list-comprehension-in-list-comprehension.